pallets/xtransfer/src/assets_wrapper.rs

```diff
@@ -8,52 +8,123 @@ pub mod pallet {
 	use frame_system::pallet_prelude::*;
 	use scale_info::TypeInfo;
 	use sp_runtime::traits::StaticLookup;
-	use sp_std::{
-		convert::{From, TryFrom},
-		result,
-	};
-	use xcm::latest::MultiLocation;
+	use sp_std::convert::From;
+	use xcm::latest::{prelude::*, MultiLocation};
 
 	#[derive(Clone, Decode, Encode, Eq, PartialEq, Ord, PartialOrd, Debug, TypeInfo)]
-	pub enum XTransferAsset {
-		ParachainAsset(MultiLocation),
-		SolochainAsset([u8; 32]),
+	pub struct XTransferAsset(MultiLocation);
+
+	impl From<MultiLocation> for XTransferAsset {
+		fn from(x: MultiLocation) -> Self {
+			XTransferAsset(x)
+		}
 	}
 
-	impl TryFrom<MultiLocation> for XTransferAsset {
-		type Error = ();
-		fn try_from(x: MultiLocation) -> result::Result<Self, ()> {
-			Ok(XTransferAsset::ParachainAsset(x))
+	impl From<XTransferAsset> for MultiLocation {
+		fn from(x: XTransferAsset) -> Self {
+			x.0
 		}
 	}
 
-	impl TryFrom<XTransferAsset> for MultiLocation {
-		type Error = ();
-		fn try_from(x: XTransferAsset) -> result::Result<Self, ()> {
-			match x {
-				XTransferAsset::ParachainAsset(location) => Ok(location),
-				_ => Err(()),
-			}
+	impl From<XTransferAsset> for [u8; 32] {
+		fn from(x: XTransferAsset) -> Self {
+			sp_io::hashing::keccak_256(&x.0.encode())
 		}
 	}
 
-	impl TryFrom<[u8; 32]> for XTransferAsset {
-		type Error = ();
-		fn try_from(x: [u8; 32]) -> result::Result<Self, ()> {
-			Ok(XTransferAsset::SolochainAsset(x))
+	pub trait AccountId32Conversion {
+		fn into_account(self) -> [u8; 32];
+	}
+
+	impl AccountId32Conversion for MultiLocation {
+		fn into_account(self) -> [u8; 32] {
+			sp_io::hashing::keccak_256(&self.encode())
 		}
 	}
 
-	impl TryFrom<XTransferAsset> for [u8; 32] {
-		type Error = ();
-		fn try_from(x: XTransferAsset) -> result::Result<Self, ()> {
-			match x {
-				XTransferAsset::SolochainAsset(rid) => Ok(rid),
-				_ => Err(()),
+	impl AccountId32Conversion for XTransferAsset {
+		fn into_account(self) -> [u8; 32] {
+			sp_io::hashing::keccak_256(&self.0.encode())
+		}
+	}
+
+	pub trait Resolve {
+		fn resolve(self) -> Option<MultiLocation>;
+	}
+
+	impl Resolve for MultiLocation {
```

Comment:
We don't care so much about the parent count, when extract the reserve location, parent count always being kept. Instead, the `first_interior` is more import. If it is `Junction::Parachain`, then we can beleive the reserve location is a parachain(If paraid is our chain, then we extract the solo chain location), otherwise, it should either be a local location or relaychain location, depends on the parent is `0` or `1` or others.

Reply:
Let's have an offline chat

Reply:
Please check the latest code